### DISCUSSION TITLE
How to get req data from nextjs middleware function to page component?

### DISCUSSION THREAD

**Discussion by siamahnaf** [vercel/next.js#43456] [Help]
I successfully write a middleware function in nextjs

```
import type { NextRequest } from 'next/server';
import { NextResponse } from 'next/server'

export function middleware(req: NextRequest) {
    const { ip, geo } = req
}

export const config = {
    matcher: '/'
}
```

Here in this middleware function, I am getting `geo` object and `ip` string. In the `geo` object it has `country`, `city`, `latitude`, `longitude`, and `region` field. Now I have to get this data into my home page component.

```
import type { NextPage, GetServerSideProps } from "next"

const Home: NextPage = () => {
  return (
    <div>
      dfd
    </div>
  );
};

export default Home;

export const getServerSideProps: GetServerSideProps = async (context) => {
  console.log(context);
  return { props: {} }
}
```

How can I get those middleware data like `ip`, `geo` in my page component.

**Comment by icyJoseph** [comment]
## Cookies

> I'd recommend this approach

From the middleware, set a cookie with the ip and geo information and read that from GSSP.

Prior to next@13, you had to set those in the response, but in next@13 you can also do it in the incoming request, https://nextjs.org/docs/advanced-features/middleware#using-cookies:

> 1. For incoming requests, cookies comes with the following methods: get, getAll, set, and delete cookies. You can check for the existence of a cookie with has or remove all cookies with clear.
> 2. For outgoing responses, cookies have the following methods get, getAll, set, and delete.

If you do set it in the response, you might want to delete it before sending the result back to the client.

## Rewriting

You could serialize somehow `ip` and `geo` and make the result part of a path, like `/pages/_private/[serializedIp]/[serializedGeo]/index.tsx`, then you could read those off the context as well, as queries IIRC.

- You'd need to forbid direct access to `/_private` from middleware
- it is rather more complex

Last but not least, an article that goes about how to work with middleware data: https://www.ctnicholas.dev/articles/how-to-use-nextjs-middleware#rewriting-pages

**Reply by icyJoseph** [reply]
I experimented a bit, and noticed that even though the `.set` API exists, for the request, it is still not going what it should...

So I reconsidered my answer to:

```ts
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  const { pathname, href } = req.nextUrl;

  console.log("Middleware", { pathname, href });

  const { ip = "127.0.0.1", geo = {} } = req;

  if (typeof ip === "undefined") {
    return NextResponse.next();
  }

  if (typeof geo === "undefined") {
    return NextResponse.next();
  }

  req.nextUrl.searchParams.set("x-ip-query", ip);

  const serializedGeo = encodeURIComponent(
    JSON.stringify(geo, (_, value) =>
      typeof value === "undefined" ? null : value
    )
  );

  req.nextUrl.searchParams.set("x-geo-query", serializedGeo);

  return NextResponse.rewrite(req.nextUrl);
}
```

So you'd rewrite for the paths that need this data.

And then in GSSP:

```tsx
import type { GetServerSidePropsContext, NextPage } from "next";
import Link from "next/link";

const GSSP: NextPage<{ geo: string; ip: string }> = ({ geo, ip }) => {
  return (
    <div>
      GSSP <Link href="/">home</Link>
      <pre>{geo}</pre>
      <p>{ip}</p>
    </div>
  );
};

export const getServerSideProps = (context: GetServerSidePropsContext) => {
  const geo = [context.query["x-geo-query"] || ""].join(""); // some trickery to please TS
  const ip = context.query["x-ip-query"] || "";

  return { props: { foo: "bar", geo: decodeURIComponent(geo), ip } };
};

export default GSSP;

```

Although... the router now see the query, so be careful with that:

![Screenshot 2022-11-28 at 16 41 13](https://user-images.githubusercontent.com/21013447/204319446-93bfdb1e-a06b-413c-a7f8-b20ef28bec78.png)

**Reply by icyJoseph** [reply]
Note that setting cookie on the response does work, but it'd require you to do a bit more of work parsing the cookie, and removing it, if you don't want to ultimately set it on the browser, which also includes making sure only the paths you care about do set it...

Posting for completeness:

```ts
import { NextRequest, NextResponse } from "next/server";

export async function middleware(req: NextRequest) {
  const { pathname, href } = req.nextUrl;

  console.log("Middleware", { pathname, href });

  const { ip = "127.0.0.1", geo = {} } = req;

  if (typeof ip === "undefined") {
    return NextResponse.next();
  }

  if (typeof geo === "undefined") {
    return NextResponse.next();
  }

  if (pathname !== "/gssp") {
   // for any other path I just continue
    return NextResponse.next();
  }
 // otherwise set cookies based off the ip and geo location
  const response = NextResponse.next();

  response.cookies.set("ip-cookie", ip); // do something similar with geo info

  return response;
}
```

And then

```tsx
import type { GetServerSidePropsContext, NextPage } from "next";
import Link from "next/link";

const GSSP: NextPage<{ geo: string; ip: string }> = ({ geo, ip }) => {
  return (
    <div>
      GSSP <Link href="/">home</Link>
      <pre>{geo}</pre>
      <p>{ip}</p>
    </div>
  );
};

export const getServerSideProps = (context: GetServerSidePropsContext) => {
  const ip = context.res.getHeader("set-cookie"); // this contains all cookies, potentially the geo cookie as well
 // https://www.npmjs.com/package/cookie might help out with the parsing

  context.res.removeHeader("set-cookie"); // avoid sending it to the client

  return { props: { foo: "bar", geo: "", ip } };
};

export default GSSP;

```

**Comment by siamahnaf** [comment]
Any code example please for doing this?
